### ISSUE TITLE
ClientFlowLoginController assumes $client->getRedirectUri() won't already have a query

### ISSUE THREAD

**Issue by RussellAult** [nextcloud/server#17885]
### Steps to reproduce
1. Setup an OAuth2 client with a "Redirection URI" that has a '?' in it (e.g. "https://foo.bar/login?component=oauth2").
2. Try to authenticate from that client.

### Expected behaviour
Nextcloud should redirect to a URI that adds the "state" key/pair with an ampersand (e.g. https://foo.bar/login?component=oauth2&state=hex).

### Actual behaviour
Nextcloud redirects to a URI that adds the "state" key/pair with a second question mark (e.g. "https://foo.bar/login?component=oauth2?state=hex").

### Work-around
In `core/Controller/ClientFlowLoginController.php` replace
```php
			$redirectUri = sprintf(
				'%s?state=%s&code=%s',
				$client->getRedirectUri(),
				urlencode($this->session->get('oauth.state')),
				urlencode($code)
			);
```

with

```php
			if (parse_url($client->getRedirectUri(), PHP_URL_QUERY)) {
				$redirectUriSeparator = '&';
			} else {
				$redirectUriSeparator = '?';
			}

			$redirectUri = sprintf(
				'%s%sstate=%s&code=%s',
				$client->getRedirectUri(),
				$redirectUriSeparator,
				urlencode($this->session->get('oauth.state')),
				urlencode($code)
			);
```

### Server configuration

**Operating system**:
Debian Buster

**Web server:**
Apache 2

**Database:**
MariaDB

**PHP version:**
7.3

**Nextcloud version:**
16.0.3

**Comment by kesselb** [comment]
Looks good to me. Mind to open a pull request with your changes?

**Comment by kesselb** [comment]
Fixed by #17893
